File: packages/sdk-js/src/inbox.ts
import { HttpClient } from './http';
import type { InboxCount, InboxItem, InboxListQuery, Paginated, RequestOptions } from './types';

export interface InboxClientOptions {
  /**
   * A short-lived END-USER token — NOT the tenant's secret API key. Mint one from your
   * backend with `NotificationClient.users.mintToken(externalUserId)` and pass it down to
   * the browser (e.g. embedded in the page, or fetched from your own authenticated endpoint).
   */
  token: string;
  baseUrl?: string;
  fetch?: typeof fetch;
  /** Per-request timeout in ms. Default 15000. */
  timeoutMs?: number;
  /** Retries beyond the initial attempt, for 429/5xx/network errors. Default 2. */
  maxRetries?: number;
}

/**
 * Browser-safe client for a single end user's in-app inbox (`/v1/inbox/*`).
 *
 * This is intentionally a separate class from `NotificationClient`: it only ever holds an
 * end-user token, so it is structurally impossible to accidentally bundle a tenant's secret
 * API key into browser-shipped code by importing the wrong client.
 *
 * Two-step flow:
 *   1. Backend (holds the secret key): `notificationClient.users.mintToken(externalUserId)`.
 *   2. Browser (holds only the minted token): `new InboxClient({ token })`.
 */
export class InboxClient {
  private readonly http: HttpClient;

  constructor(options: InboxClientOptions) {
    if (!options.token) throw new Error('token is required.');
    const fetchImpl = options.fetch ?? globalThis.fetch;
    if (!fetchImpl) throw new Error('A fetch implementation is required (pass options.fetch on runtimes without a global fetch).');
    this.http = new HttpClient({
      baseUrl: (options.baseUrl ?? 'http://localhost:3000').replace(/\/$/, ''),
      fetchImpl,
      timeoutMs: options.timeoutMs ?? 15_000,
      maxRetries: options.maxRetries ?? 2,
      authorizationHeader: `Bearer ${options.token}`,
    });
  }

  /** GET /v1/inbox */
  list(query: InboxListQuery = {}, options: RequestOptions = {}): Promise<Paginated<InboxItem>> {
    return this.http.request<Paginated<InboxItem>>('/v1/inbox', {
      query: { status: query.status, limit: query.limit, cursor: query.cursor, archived: query.archived },
      signal: options.signal,
    });
  }

  /** GET /v1/inbox/count */
  count(options: RequestOptions = {}): Promise<InboxCount> {
    return this.http.request<InboxCount>('/v1/inbox/count', { signal: options.signal });
  }

  /** POST /v1/inbox/read-all — marks every unread, unarchived item read. Safe to retry. */
  readAll(options: RequestOptions = {}): Promise<{ updated: number }> {
    return this.http.request<{ updated: number }>('/v1/inbox/read-all', { method: 'POST', idempotent: true, signal: options.signal });
  }

  /** POST /v1/inbox/:id/read — safe to retry: setting readAt twice is a no-op. */
  read(id: string, options: RequestOptions = {}): Promise<InboxItem> {
    return this.http.request<InboxItem>(`/v1/inbox/${encodeURIComponent(id)}/read`, { method: 'POST', idempotent: true, signal: options.signal });
  }

  /** POST /v1/inbox/:id/unread — safe to retry. */
  unread(id: string, options: RequestOptions = {}): Promise<InboxItem> {
    return this.http.request<InboxItem>(`/v1/inbox/${encodeURIComponent(id)}/unread`, { method: 'POST', idempotent: true, signal: options.signal });
  }

  /** POST /v1/inbox/:id/seen — safe to retry. */
  seen(id: string, options: RequestOptions = {}): Promise<InboxItem> {
    return this.http.request<InboxItem>(`/v1/inbox/${encodeURIComponent(id)}/seen`, { method: 'POST', idempotent: true, signal: options.signal });
  }

  /** POST /v1/inbox/:id/archive — safe to retry. */
  archive(id: string, options: RequestOptions = {}): Promise<InboxItem> {
    return this.http.request<InboxItem>(`/v1/inbox/${encodeURIComponent(id)}/archive`, { method: 'POST', idempotent: true, signal: options.signal });
  }
}

export function createInboxClient(options: InboxClientOptions): InboxClient {
  return new InboxClient(options);
}

File: packages/sdk-python/notification_platform/inbox.py
"""Client for a single end user's in-app inbox (``/v1/inbox/*``).

This is intentionally a separate class from ``NotificationClient``: it only ever holds a
short-lived end-user token, never the tenant's secret API key, so it is structurally
impossible to reach for the wrong credential when this is the part of your stack that
talks directly to whatever surface an end user controls.

Two-step flow:
    1. Backend (holds the secret key): ``notification_client.users.mint_token(external_user_id)``.
    2. Wherever the end user is (holds only the minted token): ``InboxClient(token=...)``.
"""

from __future__ import annotations

from typing import Any, Dict, Optional

from ._http import HttpClient


class InboxClient:
    def __init__(
        self,
        token: str,
        base_url: str = "http://localhost:3000",
        timeout_seconds: float = 15.0,
        max_retries: int = 2,
    ) -> None:
        if not token:
            raise ValueError("token is required.")
        self._http = HttpClient(
            base_url=base_url,
            authorization_header=f"Bearer {token}",
            timeout_seconds=timeout_seconds,
            max_retries=max_retries,
        )

    def list(
        self,
        status: Optional[str] = None,
        limit: Optional[int] = None,
        cursor: Optional[str] = None,
        archived: Optional[bool] = None,
    ) -> Dict[str, Any]:
        """GET /v1/inbox"""
        query: Dict[str, Any] = {"status": status, "limit": limit, "cursor": cursor}
        if archived is not None:
            query["archived"] = "true" if archived else "false"
        return self._http.request("GET", "/v1/inbox", query=query)

    def count(self) -> Dict[str, int]:
        """GET /v1/inbox/count"""
        return self._http.request("GET", "/v1/inbox/count")

    def read_all(self) -> Dict[str, int]:
        """POST /v1/inbox/read-all -- marks every unread, unarchived item read. Safe to retry."""
        return self._http.request("POST", "/v1/inbox/read-all", idempotent=True)

    def read(self, item_id: str) -> Dict[str, Any]:
        """POST /v1/inbox/:id/read -- safe to retry: setting read_at twice is a no-op."""
        return self._http.request("POST", f"/v1/inbox/{item_id}/read", idempotent=True)

    def unread(self, item_id: str) -> Dict[str, Any]:
        """POST /v1/inbox/:id/unread -- safe to retry."""
        return self._http.request("POST", f"/v1/inbox/{item_id}/unread", idempotent=True)

    def seen(self, item_id: str) -> Dict[str, Any]:
        """POST /v1/inbox/:id/seen -- safe to retry."""
        return self._http.request("POST", f"/v1/inbox/{item_id}/seen", idempotent=True)

    def archive(self, item_id: str) -> Dict[str, Any]:
        """POST /v1/inbox/:id/archive -- safe to retry."""
        return self._http.request("POST", f"/v1/inbox/{item_id}/archive", idempotent=True)
